### PR TITLE
move handling of last-seen-SYSID_MYGCS up to GCS base class

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
Needs to be split into multiple commits.

**Comment by peterbarker** [comment]
SITL-tested.

Tested that normal GCS failsafes work across Rover (including manual_control messages), Plane and Copter.

Checked that AFS works in Plane.

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/GCS_Mavlink.cpp [hunk 1]
```diff
@@ -1031,8 +1016,9 @@ void GCS_MAVLINK_Copter::handleMessage(const mavlink_message_t &msg)
         manual_override(copter.channel_throttle, packet.z, 0, 1000, tnow);
         manual_override(copter.channel_yaw, packet.r, 1000, 2000, tnow);
 
-        // a manual control message is considered to be a 'heartbeat' from the ground station for failsafe purposes
-        copter.failsafe.last_heartbeat_ms = tnow;
+        // a manual control message is considered to be a 'heartbeat'
+        // from the ground station for failsafe purposes
+        gcs().sysid_myggcs_seen(AP_HAL::millis());
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduCopter/GCS_Mavlink.cpp]: Fixed

File: ArduPlane/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -1152,17 +1152,9 @@ void GCS_MAVLINK_Plane::handleMessage(const mavlink_message_t &msg)
...
-    
-    case MAVLINK_MSG_ID_HEARTBEAT:
-    {
-        // We keep track of the last time we received a heartbeat from
-        // our GCS for failsafe purposes
-        if (msg.sysid != plane.g.sysid_my_gcs) break;
-        plane.failsafe.last_heartbeat_ms = AP_HAL::millis();
+        // a manual control message is considered to be a 'heartbeat'
+        // from the ground station for failsafe purposes
+        gcs().sysid_myggcs_seen(AP_HAL::millis());
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduPlane/GCS_Mavlink.cpp]: Fixed

File: ArduPlane/system.cpp [hunk 3]
```diff
@@ -298,6 +298,7 @@ bool Plane::set_mode_by_number(const Mode::Number new_mode_number, const ModeRea
 void Plane::check_long_failsafe()
 {
     uint32_t tnow = millis();
+    const uint32_t gcs_last_seen_ms = gcs().sysid_myggcs_last_seen_time_ms();
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduPlane/system.cpp:300]: Fixed

File: libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp [hunk 4]
```diff
@@ -204,6 +204,7 @@ AP_AdvancedFailsafe::check(uint32_t last_heartbeat_ms, bool geofence_breached, u
     }
 
     uint32_t now = AP_HAL::millis();
+    const uint32_t last_heartbeat_ms = gcs().sysid_myggcs_last_seen_time_ms();
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp:206]: Done

**Review by peterbarker — COMMENTED** [review]
All fixed.

**Review by tridge — APPROVED** [review]

**Pull request merged**